File: src/scene/LoadingScene.java
package scene;

import java.io.IOException;

import model.Condition;

import org.newdawn.slick.*;
import org.newdawn.slick.loading.DeferredResource;
import org.newdawn.slick.loading.LoadingList;
import org.newdawn.slick.state.StateBasedGame;

import component.ConditionBar;
import component.Label;
import component.Positionable.ReferencePoint;
import core.FontStore;
import core.GameDirector;
import core.ImageStore;
import core.SoundStore;

/**
 * A screen that shows a loading bar that gradually fills to 100% as
 * the game's resources are loaded (sound, images, fonts, etc).
 */
public class LoadingScene extends Scene {
	public static final SceneID ID = SceneID.LOADING;
	
	private int PADDING = 20;
	
	private int BAR_WIDTH = 200;
	private int BAR_HEIGHT = 10;
	
	private Label loadLabel;
	private Label errorLabel;
	private Condition loadCondition;
	private ConditionBar loadingBar;
	
	private DeferredResource nextResource;
	
	private String error;
	
	@Override
	public void init(GameContainer container, StateBasedGame game) throws SlickException {
		super.init(container, game);

		Font h2 = FontStore.get().getFont(FontStore.FontID.H2);
		Font field = FontStore.get().getFont(FontStore.FontID.FIELD);
		
		LoadingList.setDeferredLoading(true);

		try {
			SoundStore.get();
			ImageStore.get();
		} catch (Exception e) {
			e.printStackTrace();
			error = e.getMessage();
		}
		
		loadCondition = new Condition(0, LoadingList.get().getTotalResources(), 0); 
		
		loadingBar = new ConditionBar(container, BAR_WIDTH, BAR_HEIGHT, loadCondition);
		mainLayer.add(loadingBar, mainLayer.getPosition(ReferencePoint.CENTERCENTER), ReferencePoint.TOPCENTER);
		
		Label loadingLabel = new Label(container, h2, Color.white, "Loading...");
		mainLayer.add(loadingLabel, loadingBar.getPosition(ReferencePoint.TOPCENTER), ReferencePoint.BOTTOMCENTER, 0, -PADDING);
		
		loadLabel = new Label(container, container.getWidth(), field, Color.white, "Loading load scene...");
		mainLayer.add(loadLabel, loadingBar.getPosition(ReferencePoint.BOTTOMCENTER), ReferencePoint.TOPCENTER, 0, PADDING);
		
		errorLabel = new Label(container, container.getWidth(), field, Color.red, "Unknown Error");
		errorLabel.setVisible(false);
		mainLayer.add(errorLabel, loadLabel.getPosition(ReferencePoint.BOTTOMCENTER), ReferencePoint.TOPCENTER, 0, PADDING);
	}
	
	@Override
	public void update(GameContainer container, StateBasedGame game, int delta) throws SlickException {
		if (nextResource != null) {
			String path = nextResource.getDescription();
			
			String description = "nothing";
			if (path.contains("graphics")) {
				description = "images";
			} else if (path.contains("sounds")) {
				description = "sounds";
			}
			loadLabel.setText("Loading " + description + "...");
			
			loadCondition.increase(1);
			loadingBar.update();

			try {
				nextResource.load();
			} catch (IllegalArgumentException e) {
				e.printStackTrace();
				error = "Error: Attempted to load empty file";
			} catch (Exception e) {
				e.printStackTrace();
				error = "Error: Confused by earlier Exceptions, bailing out";
			}
			
			nextResource = null;
		}
		
		if (error != null) {
			errorLabel.setText(error);
			errorLabel.setVisible(true);
		} else {	
			if (LoadingList.get().getRemainingResources() > 0) {
				nextResource = LoadingList.get().getNext();
			} else {
				GameDirector.sharedSceneListener().requestScene(SceneID.MAINMENU, this, true);
			}
		}
	}

	@Override
	public int getID() {
		return ID.ordinal();
	}
}

File: web/src/scene/LoadingScene.ts
import { Scene } from './Scene';
import { SceneID } from './SceneID';
import { Label } from '../component/Label';
import { ConditionBar, Condition } from '../component/ConditionBar';
import { ReferencePoint } from '../component/Component';
import { Color } from '../core/Color';
import { FontStore, FontID } from '../core/FontStore';
import { ImageStore } from '../core/ImageStore';
import { SoundStore } from '../core/SoundStore';

class LoadingCondition implements Condition {
  private current: number = 0;
  private max: number = 0;

  constructor(max: number) {
    this.max = max;
  }

  getCurrent(): number {
    return this.current;
  }

  getPercentage(): number {
    if (this.max === 0) return 0;
    return this.current / this.max;
  }

  getMin(): number {
    return 0;
  }

  getMax(): number {
    return this.max;
  }

  increase(amount: number): void {
    this.current += amount;
    if (this.current > this.max) {
      this.current = this.max;
    }
  }
}

export class LoadingScene extends Scene {
  private static readonly PADDING = 20;
  private static readonly BAR_WIDTH = 200;
  private static readonly BAR_HEIGHT = 10;

  private loadLabel!: Label;
  private loadCondition!: LoadingCondition;
  private loadingBar!: ConditionBar;
  private loading: boolean = false;
  private onComplete: (() => void) | null = null;

  constructor(canvasWidth: number, canvasHeight: number, onComplete: () => void) {
    super();
    this.onComplete = onComplete;
    this.init(canvasWidth, canvasHeight);
  }

  private init(canvasWidth: number, canvasHeight: number): void {
    const h2Font = FontStore.getFont(FontID.H2);
    const fieldFont = FontStore.getFont(FontID.FIELD);

    this.loadCondition = new LoadingCondition(100);

    this.loadingBar = new ConditionBar(
      LoadingScene.BAR_WIDTH,
      LoadingScene.BAR_HEIGHT,
      this.loadCondition
    );
    this.mainLayer.add(this.loadingBar);
    this.loadingBar.setPosition(
      { x: canvasWidth / 2, y: canvasHeight / 2 },
      ReferencePoint.TOPCENTER
    );

    const loadingLabel = Label.withTextWidth(h2Font, Color.white, 'Loading...');
    this.mainLayer.add(loadingLabel);
    loadingLabel.setPosition(
      this.loadingBar.getPosition(ReferencePoint.TOPCENTER),
      ReferencePoint.BOTTOMCENTER,
      0,
      -LoadingScene.PADDING
    );

    this.loadLabel = new Label(canvasWidth, fieldFont.getLineHeight(), fieldFont, Color.white, 'Initializing...');
    this.mainLayer.add(this.loadLabel);
    this.loadLabel.setPosition(
      this.loadingBar.getPosition(ReferencePoint.BOTTOMCENTER),
      ReferencePoint.TOPCENTER,
      0,
      LoadingScene.PADDING
    );
  }

  override prepareToEnter(): void {
    super.prepareToEnter();
    this.startLoading();
  }

  private async startLoading(): Promise<void> {
    if (this.loading) return;
    this.loading = true;

    try {
      this.loadLabel.setText('Loading images...');
      await ImageStore.initialize((loaded, total) => {
        const percentage = (loaded / total) * 50;
        this.loadCondition = new LoadingCondition(100);
        this.loadCondition.increase(percentage);
        this.loadingBar.setCondition(this.loadCondition);
      });

      this.loadLabel.setText('Loading sounds...');
      await SoundStore.initialize((loaded, total) => {
        const percentage = 50 + (loaded / total) * 50;
        this.loadCondition = new LoadingCondition(100);
        this.loadCondition.increase(percentage);
        this.loadingBar.setCondition(this.loadCondition);
      });

      this.loadLabel.setText('Loading complete');
      this.loadCondition.increase(100);
      this.loadingBar.setCondition(this.loadCondition);

      setTimeout(() => {
        if (this.onComplete) {
          this.onComplete();
          this.onComplete = null;
        }
      }, 200);
    } catch (error) {
      this.loadLabel.setText('Error loading resources');
      this.loadLabel.setColor(Color.red);
      console.error('Loading error:', error);
    }
  }

  override update(delta: number): void {
    super.update(delta);
  }

  getID(): SceneID {
    return SceneID.LOADING;
  }
}
